Kafka 4.x Queue Semantics support

Shekharrajak opened the pull request:
Ref https://issues.apache.org/jira/browse/FLINK-38287

This implementation adds Kafka 4.x share group semantics to Flink's Kafka connector while maintaining full backward compatibility with existing code. The code changes are following KIP-932 and FLIP-27 main source architecture and implicit mode acknowledgement.

This directly addresses use cases where:

1. Multiple consumers need to process items efficiently in parallel from a single/multiple topic(s).
2. Messages  need explicit acknowledgment/release (to avoid reprocessing or allow retries).
Use cases where scaling Flink ML/LLM workload is critical - Shifting Kafka coordination and assignment logic to the broker side would simplify today’s complex Flink source management, making consumption more efficient, scalable, and far less error-prone.
  Operational Benefits

  - Higher Throughput: ShareGroupHeartbeat helps in Queue-like workloads, maximum throughput scenarios. Share groups distribute messages at the record level, not partition level, so multiple readers can consume from the same topic with Kafka coordinating message distribution.
  - Better Availability and Flexible Scaling: consumers assignment logic is simpler in server side and rebalancing frequency is minimised.

**Note:**

**Semantic Guarantee**
  **AT-MOST-ONCE**: Records acknowledged immediately after polling. Job failure after acknowledgment but
  before checkpoint = permanent data loss.

Ref https://issues.apache.org/jira/browse/KAFKA-19883 implemented for no data loss in commit : https://github.com/apache/flink-connector-kafka/pull/189/commits/ef83f90ca8ea1095db0883b2cc269731de0558f0

  **Good for**: Logs, monitoring, non-critical analyticsNot for: Financial transactions, critical data,
  audit trails

tomncooper committed:
[FLINK-37583] Upgrade to Kafka 4.0.0 client.

Note that this update will make the connector incompatible with Kafka
clusters running Kafka version 2.0 and older.

Signed-off-by: Thomas Cooper <code@tomcooper.dev>

Shekharrajak committed:
intial updates for kafka 4.1.0

tomncooper committed:
[FLINK-38289] Update to Flink 2.1

Shekharrajak commented on flink-connector-kafka/pom.xml:
```diff
@@ -51,7 +53,7 @@ under the License.
 			<groupId>org.apache.flink</groupId>
 			<artifactId>flink-streaming-java</artifactId>
-			<scope>provided</scope>
+			<scope>compile</scope>
```
These changes can be reverted.

Shekharrajak commented:
Using Flink SQL, we can have some validation : 

```
CREATE TABLE kafka_share_source (
      message STRING
  ) WITH (
      'connector' = 'kafka-sharegroup',
      'bootstrap.servers' = 'localhost:9092',
      'share-group-id' = 'flink-sql-test-group',
      'topic' = 'test-topic',
      'format' = 'raw',
      'source.parallelism' = '4'  -- 4 subtasks regardless of partition count
  );
 
select * from kafka_share_source;

```

<img width="2728" height="1146" alt="dafd0c9c-ec40-4af2-a929-ca27007076d8" src="https://github.com/user-attachments/assets/a65f823a-0d88-4ebe-8102-2b171ed1d690" />

jnh5y commented:
As a high-level note, since share groups do not use transactions, there will be some possibility for reprocessing messages.  Is that ok for your use cases?

Generally, do you have any performance numbers to show that this consumer is faster?  (Of course, since transactions are not available, I could imagine it being a little bit faster anyhow...)

AHeise committed:
[FLINK-38453] Add full splits to KafkaSourceEnumState

KafkaEnumerator's state contains the TopicPartitions only but not the offsets, so it doesn't contain the full split state contrary to the design intent.

There are a couple of issues with that approach. It implicitly assumes that splits are fully assigned to readers before the first checkpoint. Else the enumerator will invoke the offset initializer again on recovery from such a checkpoint leading to inconsistencies (LATEST may be initialized during the first attempt for some partitions and initialized during second attempt for others).

Through addSplitBack callback, you may also get these scenarios later for BATCH which actually leads to duplicate rows (in case of EARLIEST or SPECIFIC-OFFSETS) or data loss (in case of LATEST). Finally, it's not possible to safely use KafkaSource as part of a HybridSource because the offset initializer cannot even be recreated on recovery.

All cases are solved by also retaining the offset in the enumerator state. To that end, this commit merges the async discovery phases to immediately initialize the splits from the partitions. Any subsequent checkpoint will contain the proper start offset.

Shekharrajak committed:
major update - at most once semantics

Shekharrajak committed:
update with txn acknowledgement KAFKA-19883

Shekharrajak committed:
revert back to existing implementation: ack and commitSync

huangmingyuan commented on pom.xml:
```diff
@@ -54,7 +54,7 @@ under the License.
 		<confluent.version>7.9.2</confluent.version>
 		<flink.version>2.0.0</flink.version>
-		<kafka.version>3.9.1</kafka.version>
+		<kafka.version>4.1.0</kafka.version>
```
when will this be released?

Shekharrajak committed:
Merge branch 'apache:main' into feature/kafka4

Shekharrajak committed:
transaction manager added

Shekharrajak committed:
python api for kafka share consumer

Shekharrajak commented:
This PR served as the initial prototype that validated the core ideas behind share group integration with Flink. The code changes here proved the feasibility of checkpoint-aligned acknowledgment for at-least-once delivery with Kafka's queue semantics. I presented this work at Community Over Code and PyData Global.

Since then, the design has matured significantly. A comprehensive proposal with exhaustive failure analysis, a simplified pause/resume checkpoint model, and zero Flink side record state tracking is now documented here: https://cwiki.apache.org/confluence/x/1IEmGQ
